dodge-demo/player/Player.gd

```diff
@@ -0,0 +1,75 @@
+extends KinematicBody2D
+
+enum States { MOVING, DODGING, TAKING_DAMAGE }
+var state = States.MOVING
+
+const DIRECTION_TO_FRAME := {
+	Vector2.DOWN: 0,
+	Vector2.DOWN + Vector2.RIGHT: 1,
+	Vector2.RIGHT: 2,
+	Vector2.UP + Vector2.RIGHT: 3,
+	Vector2.UP: 4,
+}
+
+const DRAG_FACTOR := 15.0
+const MOVE_SPEED := 400.0
+const DODGE_SPEED := 1000.0
+
+var _velocity := Vector2.ZERO
+var _last_direction := Vector2.RIGHT # Assign the first direction the sprite is looking at
+
+onready var godot_sprite := $Godot
+onready var hurt_box := $HurtBox
+onready var smoke_particles := $SmokeParticles
+onready var dodge_timer := $DodgeTimer
+
+onready var move_animation_player := $MoveAnimationPlayer
+onready var damage_animation_player := $DamageAnimationPlayer
```

Comment:
Nodes should have a leading underscore if they aren't meant to be accessed from other scripts